src/common/history.c

```diff
@@ -870,48 +870,106 @@ int dt_history_copy_and_paste_on_selection(int32_t imgid, gboolean merge, GList
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+    "SELECT MAX(num) FROM main.history WHERE imgid=?1", -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
+
+  if (sqlite3_step(stmt) == SQLITE_ROW)
+    size = sqlite3_column_int(stmt, 0);
+  sqlite3_finalize(stmt);
+
+  // get history_end
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+    "SELECT history_end FROM main.images WHERE id=?1", -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
+
+  if (sqlite3_step(stmt) == SQLITE_ROW)
+    end = sqlite3_column_int(stmt, 0);
+  sqlite3_finalize(stmt);
+
+  // fprintf(stderr,"\ndt_history_end_attop for image %i: size %i, end %i",imgid,size,end);
+
+  // a special case right after removing all history
+  // It must be absolutely fresh and untouched so history_end is always on top
+  if ((size==0) && (end==0)) return -1;
+
+  // return 1 if end is larger than size
+  if (end > size) return 1;
+
+  // no compression as history_end is right in the middle of stack
+  return 0;
 }
 
+
 /* Please note: dt_history_compress_on_image 
   - is used in lighttable and darkroom mode
-  - compresses history only in the database
-  - *must* not touch the history stack
+  - It compresses history *exclusively* in the database and does *not* touch anything on the history stack
 */
 void dt_history_compress_on_image(int32_t imgid)
 {
-  // We load the specific images history if necessary
-  if(!dt_dev_is_current_image(darktable.develop, imgid))
-  {
-    darktable.develop->iop = dt_iop_load_modules(darktable.develop);
-    dt_dev_read_history_ext(darktable.develop, imgid, FALSE);
-  }
-  else
-    dt_dev_write_history(darktable.develop);
+  sqlite3_stmt *stmt;
+
+  // get history_end for image
+  int my_history_end = 0;
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+    "SELECT history_end FROM main.images WHERE id=?1", -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
+
+  if (sqlite3_step(stmt) == SQLITE_ROW)
+    my_history_end = sqlite3_column_int(stmt, 0);
+  sqlite3_finalize(stmt);
+ 
+  if (my_history_end == 0) return;
 
   // compress history, keep disabled modules as documented
-  sqlite3_stmt *stmt;
   DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
                               "DELETE FROM main.history WHERE imgid = ?1 AND num "
                               "NOT IN (SELECT MAX(num) FROM main.history WHERE "
                               "imgid = ?1 AND num < ?2 GROUP BY operation, "
                               "multi_priority)",
                               -1, &stmt, NULL);
   DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
-  DT_DEBUG_SQLITE3_BIND_INT(stmt, 2, darktable.develop->history_end);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 2, my_history_end);
   sqlite3_step(stmt);
   sqlite3_finalize(stmt);
 
-  // delete all mask_manager entries
   int masks_count = 0;
   char op_mask_manager[20] = { 0 };
+
+  bool manager_position = FALSE;
```

Comment:
bool -> gboolean

I don't remember all details (bool is C++?) but this is a style we enforce.